### ISSUE TITLE
Configuration of dual monitor setup for best usability?

### ISSUE THREAD

**Issue by smartmic** [JLErvin/berry#102]
I really like berry but I am struggling with its usability on my dual monitor setup.  I understand I can assign each workspace to either one of my two monitors.  That works, but how can I keep the windows on both monitors visible after switching workspaces?

Eg. I have open windows in workspaces 0|1 on my monitors 0|1. Works fine.
Now I switch to workspace 2 where I want to have also open windows in parallel in ws 2|3 (monitor 0|1).  Now going back to ws 0 does not show ws 1 on monitor 1.  

Usually I have a major monitor and keep auxiliary programs/windows open on my other monitor.  Switching workspaces with berry destroys the benefit of having a second monitor in my configuration.  What did I miss or what is wrong in my setup?

Is there a possibility to assign one shared workspace to two monitors (combined resolution)

Btw, there is a mistake in the [HTML documentation](https://berrywm.org/configuration.html)

> save_monitori j
> 
>     Associate the ith monitor to the jth workspace
> 

Should be _ith workspace to the jth monitor_ (man page is correct)

Here is my experimental  `autostart`:
```sh
#!/bin/bash

berryc save_monitor 0 0
berryc save_monitor 1 1
berryc save_monitor 2 0
berryc save_monitor 3 1
berryc save_monitor 4 0
berryc save_monitor 5 1
berryc save_monitor 6 0
berryc save_monitor 7 1
berryc save_monitor 8 0
berryc save_monitor 9 1

sxhkd -c ~/.config/berry/sxhkdrc &

# Set decoration geometry
berryc border_width       2
berryc inner_border_width 2
berryc title_height       20
berryc top_gap            30

# Set decoration colors
berryc focus_color         292D3E
berryc unfocus_color       292D3E
berryc inner_focus_color   FFCB6B
berryc inner_unfocus_color 565679
berryc text_focus_color    ffffff
berryc text_unfocus_color  eeeeee

# Other options
berryc smart_place "true"
berryc draw_text   "true"
berryc edge_lock   "true"
berryc set_font    "unifont"
``` 

And here my `sxhkdrc`:
```
#
# wm independent hotkeys
#

# terminal emulator
super + Return
	st -e tmux -2

# program launcher
super + d
	9menu -shell /bin/bash -file /home/martin/.config/9menu.list -popup -font "unifont"

# make sxhkd reload its configuration files:
super + Escape
	pkill -USR1 -x sxhkd

#
# berry hotkeys
#

super + {h, j, k, l}
    berryc window_move {-50 0, 0 50, 0 -50, 50 0}

super + shift +  {h, j, k, l}
    berryc window_resize {-50 0, 0 50, 0 -50, 50 0}

super + {1-9}
    berryc switch_workspace {0-8}

super + shift + {1-9}
    berryc send_to_workspace {0-8}

super + m
    berryc window_monocle

super + f
    berryc fullscreen

super +  p
    berryc snap_right

super + o 
    berryc snap_left

super + n
    berryc toggle_decorations

super + Tab
    berryc cycle_focus

super + q
    berryc window_close

super + c
    berryc window_center

~button1
    berryc pointer_focus
```

**Comment by JLErvin** [comment]
thanks for the detailed writeup. I'm looking into this...

**Comment by JLErvin** [comment]
Ok, I've found and addressed a number of issues, but I would like to test this some more before making another release. Originally, when I had added multi-monitor support, I had tested this behavior. But it seems between patches and my lack of multiple monitors for proper testing, proper functionality was lost.

The first issue involves workspace switching which hides clients on a workspace but should not. For a clear example, suppose workspace 1 and 3 are mapped to monitor 0, while workspace 2 and 4 are mapped to monitor 1. With clients open on both workspace 1 and 3, switching to workspace  2 should _not_  hide the contents of workspace 3. This behavior has been correctly re-added as of fd8e2bc.

The second issue involves sending clients to workspaces which are already mapped but not focused. As per the mapping in the previous example, suppose workspaces 1 and 3 are mapped on monitors 0 and 1, respectively. A new window is spawned on workspace 1 and send to workspace 3. This new window should be immediately visible on workspace 3, since it is currently mapped on monitor 1. This behavior has been correctly re-added as of fd8e2bc.

The final issue involves mapping of new windows on multiple monitor setups. Users with `client_place` enabled would see newly spawned windows not correctly mapping to non-primary monitors. This has been fixed as of 
0368e8b.

I'm going to take a day or two more to do some more testing, but in the meantime you can (and it would make me so so happy if you would) test out these changes on the development branch [huckleberry](https://github.com/JLErvin/berry/tree/huckleberry). Let me know if you have any questions. Cheers @smartmic

**Comment by smartmic** [comment]
@JLErvin thanks for the quick fixes on this issue, I'll give my best to support you on testing.  Here are my comments/findings so far:

> The first issue involves workspace switching which hides clients on a workspace but should not. For a clear example, suppose workspace 1 and 3 are mapped to monitor 0, while workspace 2 and 4 are mapped to monitor 1. With clients open on both workspace 1 and 3, switching to workspace 2 should not hide the contents of workspace ~~3~~ **1**. This behavior has been correctly re-added as of fd8e2bc.

✔ Confirmed (using … *not hide the contents of workspace 1* though). I can now see content in workspace combinations 1|2, 1|4, 3|2, 3|4 on both monitors (1,3 - left monitor, 2,4 - right monitor) - as expected.

> The second issue involves sending clients to workspaces which are already mapped but not focused. As per the mapping in the previous example, suppose workspaces 1 and 3 are mapped on monitors 0 and 1, respectively. A new window is spawned on workspace 1 and send to workspace 3. This new window should be immediately visible on workspace 3, since it is currently mapped on monitor 1. This behavior has been correctly re-added as of fd8e2bc.

✘ Not yet 100% working for me, there is still one unexpected behavior, at least in my setup: I am in workspace 1 on monitor 1 and open a new window (e.g. xeyes). When I send it to workspace 3, the window *jumps to the right edge of my monitor 1*, instead of going out of sight in current workspace 1.  When I switch to workspace 3 though, the new window is there. Going back to workspace 1 does not show the window there, so the initial "send" anomaly  is gone.

> The final issue involves mapping of new windows on multiple monitor setups. Users with client_place enabled would see newly spawned windows not correctly mapping to non-primary monitors. This has been fixed as of
0368e8b.

Hm, not shure how I can test this. What is `client_place`, I have not used this command, is it documented?

**Comment by JLErvin** [comment]
Ok, with regards to the second issue I believe I have this fixed. I also implemented it such that sending windows between monitors _should_ maintain their relative position on the previous monitor. For example, if a window is on the middle right half of the screen and you send it to the next monitor, it's position will be maintained on the right half of the destination monitor. There are some weird cases where the resolutions differ, in which case a window may be force to the edge of the next monitor, but I'm ok with that. Go ahead and pull the latest development branch and see if it fixes what you were talking about.

As for point 3, this isn't something the user does (I probably should have specified this). `client_place` is the function that implements `berry`'s smart window placement algorithm. Basically, windows that appear on new monitors should be placed in unused spaces. You shouldn't have to worry about this one as I think it works ok with my testing. If you're curious, the `berryc` function `smart_place` toggles this behavior. With it disabled, new windows will always be spawned in the middle of your screen. With it enabled, new windows will be placed such that they don't overlap with existing windows (if  possible).

**Comment by JLErvin** [comment]
After some usage the functionality appears ok to me - I'm going to merge with master. We can open a new ticket and cut a new release if there are additional issues. Thanks for the help @smartmic :)

**Comment by smartmic** [comment]
No problem, happy I could help. And I can also confirm the new usability - way better than before and a big thumbs up for berry!
